### PR TITLE
Support tree-rollout

### PR DESCRIPTION
**Pull request by li2zhi**
# PR type
- [x] New Feature

# PR information
Inspired by [TreePO](https://github.com/multimodal-art-projection/TreePO), we implemented multi-turn tree-rollout based on GRPO, reducing inference overhead and improving training efficiency.

In GRPO-based training, the efficiency of rollouts is critical. The current multi-turn reasoning workflow in Swift is fully streaming-based, which means several key steps are repeatedly executed across generations without reuse. This results in unnecessary duplicated computation and significantly increases inference overhead.

To address this issue, we drew inspiration from [TreePO](https://github.com/multimodal-art-projection/TreePO) and reformulated the entire sampling phase as a tree-search process. In this design, num_generations is interpreted as the tree’s maximum width, while max_turns represents its maximum depth. Before each reasoning step, the scheduler dynamically determines whether the current node should expand and how many branches it should allocate, based on the current tree width and the branching strategy.

Furthermore, when reasoning terminates early and the number of leaf nodes has not yet reached the maximum width, a backtracking mechanism is activated. The scheduler selects earlier nodes for additional branch expansion, allowing the rollout to continue until the desired sample count is fulfilled.

This tree-rollout approach reuses intermediate computation, reduces redundant inference, and improves overall training efficiency.

tree-rollout

Related to issue: [Support TreePO #5708](https://github.com/modelscope/ms-swift/issues/5708)

## Experiment results
Experiment results

### CODE REVIEW

File: swift/trainers/rlhf_trainer/rollout_mixin.py [hunk 1]
```diff
@@ -724,8 +725,12 @@ def _server_rollout(self, inputs: DataType, request_config: RequestConfig,
...
 
         if self.accelerator.is_main_process:
-            all_outputs: List[RolloutOutput] = self._engine_infer(
-                infer_requests=all_requests, request_config=request_config)
+            if self.args.tree_rollout:
+                all_outputs: List[RolloutOutput] = self.multi_turn_scheduler.run(
+                    infer_request=all_requests, request_config=request_config)
+            else:
+                all_outputs: List[RolloutOutput] = self._engine_infer(
+                    infer_requests=all_requests, request_config=request_config)
```
Comment (hjh0119) [inline swift/trainers/rlhf_trainer/rollout_mixin.py]: scheduler.run is executed in server mode with async engine, 

For colocate, async engine will be supported soon, so we may not need to modify this part now.

File: swift/trainers/rlhf_trainer/rollout_mixin.py [hunk 2]
```diff
@@ -137,7 +138,7 @@ def _prepare_vllm(self):
                 self.vllm_client.get_engine_type()
                 vllm_use_async_engine = [self.vllm_client.use_async_engine]
                 use_gym_env = [self.vllm_client.use_gym_env]
-                enable_multi_turn = [self.vllm_client.enable_multi_turn]
+                enable_multi_turn = [self.vllm_client.enable_multi_turn or args.tree_rollout]
```
Comment (hjh0119) [inline swift/trainers/rlhf_trainer/rollout_mixin.py]: See the reasons below.

File: swift/trainers/rlhf_trainer/rollout_mixin.py [hunk 3]
```diff
@@ -108,6 +108,7 @@ def _prepare_rollout_params(self):
             top_k=args.top_k,
             repetition_penalty=args.repetition_penalty,
             stop=args.stop_words,
+            logprobs=args.tree_rollout,
```
Comment (hjh0119) [inline swift/trainers/rlhf_trainer/rollout_mixin.py]: can we set it in scheduler.run?

File: swift/plugin/multi_turn.py [hunk 4]
```diff
@@ -662,8 +668,186 @@ async def run(self, infer_request: 'RolloutInferRequest', request_config: 'Reque
...
+    """
+    def __init__(self, vllm_client: VLLMClient = None, *args, **kwargs):
+        self.max_tree_width = kwargs['args'].num_generations
+        self.max_tree_deep = kwargs['args'].max_tree_deep
+        self.max_divergence = kwargs['args'].tree_max_divergence
+        self.divergence_strategy = kwargs['args'].tree_divergence_strategy
+        self.root_divergence = kwargs['args'].tree_root_divergence
+
+        self.vllm_client = vllm_client
+        self.executor = ThreadPoolExecutor(max_workers=self.max_tree_width)
```
Comment (hjh0119) [inline swift/plugin/multi_turn.py]: Could you please remove it to example directory, just as [deepeyes](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/deepeyes) does?

File: examples/train/grpo/plugin/treepo/tree_rollout.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    messages: Messages
+    logprobs: List[List[float]] = field(default_factory=list)
+
+    all_response_ids: List[List[int]] = field(default_factory=list)
+    last_response: ChatCompletionResponseChoice = None
+
+    token_count_per_step: List[int] = field(default_factory=list)
+
+    status: SampleStatus = SampleStatus.INITIAL
```
Comment (hjh0119) [inline examples/train/grpo/plugin/treepo/tree_rollout.py:48]: same as scheduler, plz remove to example directory

File: swift/trainers/rlhf_trainer/rollout_mixin.py [hunk 6]
```diff
@@ -933,7 +938,13 @@ def _prepare_scheduler(self):
...
             if isinstance(args.multi_turn_scheduler, str):
                 assert args.multi_turn_scheduler in multi_turns
-                multi_turn_scheduler = multi_turns[args.multi_turn_scheduler](max_turns=args.max_turns)
+                multi_turn_scheduler_class = multi_turns[args.multi_turn_scheduler]
+
+                if args.tree_rollout:
+                    assert issubclass(multi_turn_scheduler_class, TreeRolloutScheduler)
+
+                multi_turn_scheduler = multi_turn_scheduler_class(
+                    max_turns=args.max_turns, vllm_client=self.vllm_client, args=args)
```
Comment (hjh0119) [inline swift/trainers/rlhf_trainer/rollout_mixin.py]: Same here — please revert it and keep the relevant arguments within the scheduler itself.

File: swift/trainers/arguments.py [hunk 7]
```diff
@@ -309,6 +309,13 @@ class GRPOArgumentsMixin(RolloutTrainerArgumentsMixin):
     completion_length_limit_scope: Literal['total', 'per_round'] = 'per_round'
     vllm_server_pass_dataset: bool = False
 
+    # tree rollout
+    tree_rollout: bool = False
+    max_tree_deep: Optional[int] = 8
+    tree_root_divergence: Optional[int] = 1
+    tree_max_divergence: Optional[int] = 2
+    tree_divergence_strategy: Optional[str] = 'logprob'
+
```
Comment (hjh0119) [inline swift/trainers/arguments.py]: please keep the relevant arguments within the scheduler itself.

File: examples/train/grpo/external/vllm_tree_rollout.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    --top_p 0.9 \
+    --top_k 50 \
+    --deepspeed zero2 \
+    --gradient_checkpointing false \
+    --log_completions true \
+    --num_iterations 1 \
+    --beta 0.04 \
+    --tree_rollout true \
+    --multi_turn_scheduler tree_rollout_scheduler \
+    --max_tree_deep 4
```
Comment (hjh0119) [inline examples/train/grpo/external/vllm_tree_rollout.sh]: Please create a new directory to include the script and related plugin, making it look more complete, similar to the [deepeyes](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/deepeyes) example.

**Review by hjh0119 — CHANGES_REQUESTED** [review]

**Comment by hjh0119** [comment]
Thanks for your contribution! I've left a few some revision suggestions.

Meanwhile, the following additions would be great:

1. Experimental results comparing with the original implementation.
2. Relevant papers or technical documentation in https://github.com/modelscope/ms-swift/tree/main/docs/source/Instruction/GRPO/AdvancedResearch, or best practices documentation in https://github.com/modelscope/ms-swift/tree/main/docs/source/BestPractices.

File: swift/trainers/rlhf_trainer/rollout_mixin.py (hunk 1 above)
Comment (li2zhi) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: Thanks for the suggestion! I have one point I’d like to clarify.

scheduler.run defines the full multi_turn_infer workflow, but from what I can see in the current codebase, this method is not actually used. I also noticed that _colocate_multi_turn_infer is implemented in rollout_mixin, but tree-rollout cannot reuse that logic.

If we remove this branch now, how can we ensure that the custom multi-turn logic for tree-rollout scheduler will still be executed?

Comment (hjh0119) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: https://github.com/modelscope/ms-swift/issues/6328

Comment (li2zhi) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: got it, thanks a lot

File: swift/trainers/rlhf_trainer/rollout_mixin.py (hunk 3 above)
Comment (hjh0119) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: FYI, logprobs will be set to true in https://github.com/modelscope/ms-swift/pull/6678 if args.use_vllm true

File: swift/trainers/rlhf_trainer/rollout_mixin.py (hunk 6 above)
Comment (y2logic) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: > Same here — please revert it and keep the relevant arguments within the scheduler itself.

There is a bit of a conflict here.
GRPO training uses RepeatSampler to replicate samples for rollouts (mini_repeat_count = self.num_generations). However, this replication is unnecessary in tree rollouts, since the branching naturally produces more trajectories and reduces inference overhead.

Comment (hjh0119) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: idk where the special modification is here. 

If it's about changes to the request config or infer request, I think it could be handled within the scheduler.

Comment (y2logic) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]:
```python
    def _get_train_sampler(self, train_dataset=None):
        if self.template.sequence_parallel_size > 1:
            from swift.trainers.sequence_parallel import sequence_parallel
            if self.args.tree_rollout:
                mini_repeat_count = 1
            else:
                mini_repeat_count = self.num_generations

            return RepeatSampler(
                data_source=train_dataset or self.train_dataset,
                mini_repeat_count=mini_repeat_count,
                batch_size=self.args.generation_batch_size // self.num_generations,
                repeat_count=self.num_iterations * self.args.steps_per_generation * sequence_parallel.world_size,
                shuffle=self.shuffle_dataset,
                seed=self.args.seed,
            )
        else:
            return super()._get_train_sampler(train_dataset)
```
It's about the RepeatSampler configuration.

Comment (hjh0119) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: RepeatSampler is designed to repeat N times for infer request, can we post-process the infer requests within scheduler?

Comment (y2logic) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: Actually we tried to deduplicate the infer request, but it won't work if DP enabled in rollout stage.

Comment (hjh0119) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: If `mini_repeat_count` is set to 1, each GPU's `infer_request` would change from `(per_device_batch_size)` completion requests to `(per_device_batch_size)` prompt requests, which might require modifications in many more places.

> But it won't work if DP (Data Parallelism) is enabled during the rollout stage.

This issue should only occur when a single prompt is split across multiple devices. Maybe we can avoid this by explicitly stating it in the documentation or script.

File: docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+## Implementation Details
+[TreePO implementation example](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/treepo/tree_rollout_plugin.py), which references the [official implementation](https://github.com/multimodal-art-projection/TreePO/blob/main/recipe/treepo/vllm_rollout_tree.py) provides sample code for a TreePO training plugin，covering logic related to multi-round interactions, termination judgment, and branch rollback.
+
+**Note:** In actual use, you need to rewrite the logic of methods such as step and check_finished according to your own scenario requirements to ensure that they can execute as expected in the custom scenario.
+
+The complete training script can be found at [script](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/treepo/tree_rollout.sh).
+
+## Test Data
+| /                       | batch size | num generation | max_tree_deep | total_infer | saving ratio |
```
Comment (hjh0119) [inline docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md]: Does "total_infer" refer to the rollout time? Is the unit in milliseconds (ms)?

File: swift/trainers/rlhf_trainer/rollout_mixin.py (hunk 6 above)
Comment (li2zhi) [reply swift/trainers/rlhf_trainer/rollout_mixin.py]: The relevant code has been adjusted, and all logic is now within the scheduler.

File: docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md (hunk 9 above)
Comment (y2logic) [reply docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md]: Sorry, this phrasing is a bit confusing. `total_infer` refers to the number of inference calls, not the rollout running time.

Comment (hjh0119) [reply docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md]: Can we add some descriptions to clarify it?

Comment (li2zhi) [reply docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md]: Thanks, we've added the descriptions.

File: docs/source/Instruction/GRPO/AdvancedResearch/treepo.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+训练参考该[脚本](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/treepo/tree_rollout.sh)
+
+## 测试数据
+> model: Qwen/Qwen2.5-0.5B
+> dataset: AI-MO/NuminaMath-TIR
+> subset size: 1,000 samples
+> 1 GPU for training, 1 GPU for inference
+
+| \                       | batch_size | num_generation | max_tree_deep | global_step | total inference calls | saving ratio | train_speed(iter/s) | improvement rate |
```
Comment (hjh0119) [inline docs/source/Instruction/GRPO/AdvancedResearch/treepo.md]:
```suggestion
| \                       | batch_size | num_generation | max_tree_depth | global_step | total inference calls | saving ratio | train_speed(iter/s) | improvement rate |
```

File: docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+The complete training script can be found at [script](https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo/plugin/treepo/tree_rollout.sh).
+
+## Test Data
+> model: Qwen/Qwen2.5-0.5B
+> dataset: AI-MO/NuminaMath-TIR
+> subset size: 1,000 samples
+> 1 GPU for training, 1 GPU for inference
+
+| \                       | batch_size | num_generation | max_tree_deep | global_step | total inference calls | saving ratio | train_speed(iter/s) | improvement rate |
```
Comment (hjh0119) [inline docs/source_en/Instruction/GRPO/AdvancedResearch/treepo.md]:
```suggestion
| \                       | batch_size | num_generation | max_tree_depth | global_step | total inference calls | saving ratio | train_speed(iter/s) | improvement rate |
```

File: examples/train/grpo/plugin/treepo/tree_rollout_plugin.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    CUSTOMIZATION:
+        Implement the required `step()` method and optionally override `check_finished()`
+        - Uses TreeRolloutScheduler's run() method infrastructure
+        - Only need to implement turn transition logic in step()
+        - Optionally customize termination conditions
+
+    Attributes:
+        max_tree_width (int):
+            For GRPO, it must be equal to num_generations.
+        max_tree_deep (int):
```
Comment (hjh0119) [inline examples/train/grpo/plugin/treepo/tree_rollout_plugin.py]:
```suggestion
        max_tree_depth (int):
```

**Comment by hjh0119** [comment]
LGTM
thanks!
